File: src/go/problems/leaf_similar_test.go
// https://leetcode.com/problems/leaf-similar-trees

package problems

import (
	"testing"
)

func TestLeafSimilar(t *testing.T) {
	var node1 = TreeNode{Val: 3, Left: &TreeNode{Val: 5, Left: &TreeNode{Val: 6}, Right: &TreeNode{Val: 2, Left: &TreeNode{Val: 7}, Right: &TreeNode{Val: 4}}}, Right: &TreeNode{Val: 1, Left: &TreeNode{Val: 9}, Right: &TreeNode{Val: 8}}}
	var node2 = TreeNode{Val: 3, Left: &TreeNode{Val: 5, Left: &TreeNode{Val: 6}, Right: &TreeNode{Val: 7}}, Right: &TreeNode{Val: 1, Left: &TreeNode{Val: 4}, Right: &TreeNode{Val: 2, Left: &TreeNode{Val: 9}, Right: &TreeNode{Val: 8}}}}
	
	result := LeafSimilar(&node1, &node2)
	t.Log(result)
}

func LeafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	var dfs func (node *TreeNode, result *[]int)

	dfs = func (node *TreeNode, result *[]int)  {
		if node != nil && node.Left == nil && node.Right == nil {
			*result = append(*result, node.Val)
			return
		}
	
		if node.Left != nil {
			dfs(node.Left, result)
		}

		if node.Right != nil {
			dfs(node.Right, result)
		}
	}

	result1 := []int{}	
	dfs(root1, &result1)

	result2 := []int{}
	dfs(root2, &result2)

	if len(result1) != len(result2) {
		return false
	}

	for i := 0; i < len(result1); i++ {
		if result1[i] != result2[i] {
			return false
		}
	}

	return true
}

File: src/python/problems/leaf_similar_test.py
# https://leetcode.com/problems/leaf-similar-trees/

from typing import Optional
from core.problem_base import ProblemBase, TestGen
from models.tree_node import TreeNode

class LeafSimilar(ProblemBase):
    def Solution(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def Dfs(node, result) -> int:
            if node and not node.left and not node.right:
                result.append(node.val)
                return 
            
            if node.left:
                Dfs(node.left, result)
            
            if node.right:
                Dfs(node.right, result)

        result1 = []
        Dfs(root1, result1)

        result2 = []
        Dfs(root2, result2)

        return result1 == result2

if __name__ == '__main__':
    node1 = TreeNode(3, TreeNode(5, TreeNode(6), TreeNode(2, TreeNode(7), TreeNode(4))), TreeNode(1, TreeNode(9), TreeNode(8)))
    node2 = TreeNode(3, TreeNode(5, TreeNode(6), TreeNode(7)), TreeNode(1, TreeNode(4), TreeNode(2, TreeNode(9), TreeNode(8))))

    TestGen(LeafSimilar) \
        .Add(lambda tc: tc.Param("root1", node1).Param("root2", node2).Result(True)) \
        .Run()